landlab/bmi/bmi_bridge.py

```diff
@@ -310,42 +313,88 @@ def get_value_ref(self, name):
             """Get a copy of a variable's data."""
-            return self._base.grid.at_node[name].copy()
+            dest[:] = self._base.grid.at_node[name]
+            return dest
 
-        def set_value(self, name, vals):
+        def set_value(self, name, values):
             """Set the values of a variable."""
             if name in self.get_input_var_names():
                 if name in self._base.grid.at_node:
-                    self._base.grid.at_node[name][:] = vals.flat
+                    self._base.grid.at_node[name][:] = values.flat
                 else:
-                    self._base.grid.at_node[name] = vals
+                    self._base.grid.at_node[name] = values
             else:
                 raise KeyError("{name} is not an input item".format(name=name))
 
-        def get_grid_origin(self, gid):
+        def get_grid_origin(self, grid, origin):
             """Get the origin for a structured grid."""
-            return (self._base.grid.node_y[0], self._base.grid.node_x[0])
+            origin[:] = (self._base.grid.node_y[0], self._base.grid.node_x[0])
+            return origin
 
-        def get_grid_rank(self, gid):
+        def get_grid_rank(self, grid):
             """Get the number of dimensions of a grid."""
             return 2
 
-        def get_grid_shape(self, gid):
+        def get_grid_shape(self, grid, shape):
             """Get the shape of a structured grid."""
-            return (
+            shape[:] = (
                 self._base.grid.number_of_node_rows,
                 self._base.grid.number_of_node_columns,
             )
+            return shape
 
-        def get_grid_spacing(self, gid):
+        def get_grid_spacing(self, grid, spacing):
             """Get the row and column spacing of a structured grid."""
-            return (self._base.grid.dy, self._base.grid.dx)
+            spacing[:] = (self._base.grid.dy, self._base.grid.dx)
+            return spacing
 
-        def get_grid_type(self, gid):
+        def get_grid_type(self, grid):
             """Get the type of grid."""
             return "uniform_rectilinear"
 
+        def get_grid_edge_count(self, grid):
+            return self._base.grid.number_of_links
+
+        def get_grid_edge_nodes(self, grid, edge_nodes):
+            return self._base.grid.nodes_at_link.reshape((-1,))
+
+        def get_grid_face_count(self, grid):
+            return self._base.grid.number_of_patches
+
+        def get_grid_face_nodes(self, grid, face_nodes):
+            return self._base.grid.nodes_at_patch
+
+        def get_grid_node_count(self, grid):
+            return self._base.grid.number_of_nodes
+
+        def get_grid_nodes_per_face(self, grid, nodes_per_face):
+            raise NotImplementedError("get_grid_nodes_per_face")
+
+        def get_grid_size(self, grid):
+            return self._base.grid.number_of_nodes
+
+        def get_grid_x(self, grid, x):
+            return self._base.grid.x_of_node
```

Comment:
A general BMI question. Since uniform rectilinear grids don't _need_ `get_grid_x`.... Is it a problem that it is implemented? Or is it OK?

Ultimately this is useful for making the BMI grid functions cross-grid compatible... That is, if it is OK to have `get_grid_x`, etc for RasterModelGrids, then we can just raise NotImplemented errors after testing for Hex/Radial/Voronoi in the appropriate functions (e.g., get grid shape).

Reply:
No, I don't think it's a problem that `get_grid_x` is defined for a uniform rectilinear grid. It's probably just redundant and should never get called. For other grid types we'll use it.